README.md
# PyMamba

[![Build Status](https://travis-ci.org/oddjobz/pymamba.svg?branch=master&v=17)](https://travis-ci.org/oddjobz/pymamba)
[![Coverage](https://codecov.io/gh/oddjobz/pymamba/branch/master/graph/badge.svg)](https://codecov.io/gh/oddjobz/pymamba)
[![PyPI version](https://badge.fury.io/py/pymamba.svg)](https://badge.fury.io/py/pymamba)
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=Y8KQE7TRJV6ZA)

PyMamba is a Python library designed to turn the LMDB storage engine into a relatively 
complete Database package for use in Python3 applications. In the spirit of the 
language title, this package is named after the fastest snake on the planet 
(*Black Mamba*) as a nod to the speed of the LMDB storage engine we're leaning on.

The logic behind a *Python* only database is simply that most modern applications 
tend to be written in higher level languages such as Python, and access speeds when 
measured from the likes of 'C' are in context irrelevant. The driving force behind 
this development was my conclusion that using MySQL for web based applications is 
outdated, and the performance of my preferred alternative (Mongo) is really pretty 
poor. When I stopped to think in more depth, the Python MySQL interface leaves things 
to be desired, and the Mongo interface is primarily JS orientated rather than Pythonic,
so definitely a gap in the market so to speak.

Anyway, without wanting to bore you, PyMamba should be noticeably faster than Mongo, 
and considerably ligther in terms of resource usage. (*so by implication far easier 
to deploy and maintain*)

#### PyMamba ORM (New!) :information_source:

There is now a native ORM style interface now available. It's very new and there are
some features that haven't yet been implemented, so be gentle with it. There's some
initial documentation here;
* [README-ORM.md](https://oddjobz.github.io/pymamba/README-ORM.html)

###### Useful links to other documentation:
* [Sphinx API documentation](https://pymamba.linux.co.uk)
* [A hisory of all Changes](https://oddjobz.github.io/pymamba/CHANGELOG.html)
* [PyMamba on 'Github pages'](https://oddjobz.github.io/pymamba)

### Features

* LMDB mmapped storage engine, arguably the *best* solution for small to medium DB's
* Data is mapped directly to Python objects (ujson) so you read and write *dict's*
* Transparent indexes are keyed by Python 'expression'
* There is no schema, DDL, query language, just *Python*
* Everything is transactional with ACI[D] by default
* LMDB provides multi-thread and multi-process access which defeats GIL limitations

### Examples

This is an example of how to create a new database called *my-database*, then within 
that database to create a table called *people*, then to add some people. 
(this is all *from* the Python shell)

```python
from pymamba import Database
db = Database('my-database')
people = db.table('people')
people.append({'name': 'Fred Bloggs', 'age': 21})
people.append({'name': 'Joe Smith', 'age': 22})
people.append({'name': 'John Doe', 'age': 19})
```
Now there are lots of different ways of recovering information from the database, 
the simplest is just to use **find()** which can be used to scan through the entire 
table. As find returns a generator, you can either use it within a for-loop, or use 
*list* to recover the results as a single list object.

```python
>>> list(people.find())
[{'_id': b'58ed69161839fc5e5a57bc35', 'name': 'Fred Bloggs', 'age': 21}, {'_id': b'58ed69211839fc5e5a57bc36', 'name': 'Joe Smith', 'age': 22}, {'_id': b'58ed69301839fc5e5a57bc37', 'name': 'John Doe', 'age': 19}]

>>> for doc in people.find():
...     print(doc)
... 
{'_id': b'58ed69161839fc5e5a57bc35', 'name': 'Fred Bloggs', 'age': 21}
{'_id': b'58ed69211839fc5e5a57bc36', 'name': 'Joe Smith', 'age': 22}
{'_id': b'58ed69301839fc5e5a57bc37', 'name': 'John Doe', 'age': 19}

```
Note that the returned record includes an *_id* field, this is almost identical to 
the ObjectId field used by Mongo, except we're returning a simple byte-string rather 
than an ObjectId class. A nice feature of dealing with data in this form when matched 
with Python's new 'format' function is the ability to easily format this data like so;

```python
>>> for doc in people.find():
...     print('Name: {name:20} Age:{age:3}'.format(**doc))
... 
Name: Fred Bloggs          Age: 21
Name: Joe Smith            Age: 22
Name: John Doe             Age: 19
```
Or if we just want a subset of the data, we can use an anonymous function to filter our results;
(*note that this is a linear / sequential scan with a filter*)
```python
>>> for doc in people.find(expression=lambda doc: doc['age'] > 21):
...     print('Name: {name:20} Age:{age:3}'.format(**doc))
... 
Name: Joe Smith            Age: 22

```


### Indexing
Transparent indexes are a key part of any database system, and I struggled for a 
while trying to decide which mechanism to use. On the one hand I wanted the 
functionality of being able to index tables by compound fields and functions, and on 
the other I just wanted to be able to simply index based on a single clean field. 
In the end I settled on the following;

```python
>>> people.index('by_name', '{name}')
>>> people.index('by_age_name', '{age:03}{name}')
```
If you're really familiar with Python format strings, you're going to see fairly 
quickly what's going on here, essentially we're indexing by expression only, but 
the expression comes from a Python format string when supplied with the record in 
*dict* format. So you can't directly use a function to do *anything* with regards 
to key generation, but you can do an awful lot with the 
Python [format *mini-language*](https://docs.python.org/3.4/library/string.html#formatspec). 
(and adding actual functions is relatively easy for anyone who can think of a must-have use-case)

So, once we have an index we can search using the index and also find records in 
order based on the index, so we can re-use *find* but this time give it an index to use;
```python
>>> for doc in people.find('by_age_name'):
...     print('Name: {name:20} Age:{age:3}'.format(**doc))
... 
Name: John Doe             Age: 19
Name: Fred Bloggs          Age: 21
Name: Joe Smith            Age: 22
```
Or we can look for specific records;
```python
>>> people.seek_one('by_name', {'name': 'Joe Smith'})
{'_id': b'58ed69211839fc5e5a57bc36', 'name': 'Joe Smith', 'age': 22}
```
Or we can look for a range of records;
```python
>>> for doc in people.range('by_name', {'name': 'J'}, {'name': 'K'}):
...     print('Name: {name:20} Age:{age:3}'.format(**doc))
... 
Name: Joe Smith            Age: 22
Name: John Doe             Age: 19
```
### Updating Records
We've already covered adding new records to the database, so that leaves us with
updating and deleting records. How about this;
```python
>>> person = people.seek_one('by_name', {'name': 'Joe Smith'})
>>> person['age'] += 1
>>> people.save(person)
>>> people.seek_one('by_name', {'name': 'Joe Smith'})
{'_id': b'58ed69211839fc5e5a57bc36', 'name': 'Joe Smith', 'age': 23}
```
And to delete;
```python
>>> person = people.seek_one('by_name', {'name': 'Fred Bloggs'})
>>> people.delete(person['_id'])
>>> for doc in people.find():
...     print('Name: {name:20} Age:{age:3}'.format(**doc))
... 
Name: Joe Smith            Age: 23
Name: John Doe             Age: 19
>>> 
```
There's a lot more to come, but so far it's looking pretty promising.
On my workstation a for-loop based on a **find** yields around **200k** results per second, and an **append** yields around 30k new items per second. This seems to be fairly respectable for a high level language database and seems to be much faster than Mongo when used with either Python or Node.

```text
** SINGLE Threaded benchmark **
** Probably better throughput with multiple processes

* No Indecies
  -     0: 5000 - Append Speed/sec = 48882
  -  5000: 5000 - Append Speed/sec = 52778
  - 10000: 5000 - Append Speed/sec = 52882
* Indexed by sid, day, hour
  -     0: 5000 - Append Speed/sec = 34420
  -  5000: 5000 - Append Speed/sec = 36096
  - 10000: 5000 - Append Speed/sec = 35885
* Indexed by function
  -     0: 5000 - Append Speed/sec = 39235
  -  5000: 5000 - Append Speed/sec = 39822
  - 10000: 5000 - Append Speed/sec = 41116
* Linear scan through most recent index
  -     0:15000 - Read Speed/sec   = 234615
```
[![paypal](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=Y8KQE7TRJV6ZA)



CHANGELOG.md
### Version 0.3.0

* Fixed bug in 'drop' that prevented access to system tables
* Added more unit testing for 'drop'
* Restructured how calculated values are stored. Putting them in the root
  folder with table definitions was causing problems for mdb_copy so they now
  reside in their own folder called __metadata__. Pre-existing items will be
  transparently migrated on startup.

### Version 0.2.9

* Updating README.rst

### Version 0.2.8

* Changed requirements, backed out of previous change

### Version 0.2.7

* Changed requirements, replaced pymongo with bson

### Version 0.2.6

* Additional unit testing (address_book)
* some bug fixes in relation to address_book

### Version 0.2.5

* Couple of critical bug fixes in the new ORM code
* Couple of related fixes to the addressbook demo
* A little refactoring

### Version 0.2.4

* Adding addressbook demo (nested relationships)
* A number of related bugfixes

### Version 0.2

* Complete restructure to allow for application level transactions (for replication)
* Privatised some methods that don't need to be exposed
* Database now has a **.binlog(bool)** function (logging off by default)
* Database now has **.begin()** and **.end()** methods (begin supports **with**)
* All **Table** functions should now be wrapped to ensure they're included in repl transactions
* ORM - we now have a built-in (optional) ORM with support for custom models and custom types
* ORM - includes support for calculated (virtual) fields and transparent many-to-many links
* Added README-ORM
* Main test suite was added to and transaction testing added
* Two new suites were added for model and ORM testing
* Tidied up some of the examples and benchmark
* Updated README to include link to ORM docs and sample benchmark results

### Version 0.1.35

* Fixed bug in 'seek' where the index key rather than the primary key was returned in the record body
* Clean up exception handling on append
* Fix append to allow the use of user-generated _id's
* Facilitate environment-wide txn

### Version 0.1.32

* Updated range code to include None as a possible upper and/or lower limit

### Version 0.1.31

* Added three new test cases
* Added "inclusive" option to table.range, set False to exclude matching keys at 
each end of the range results.

### Version 0.1.30

* _id no longer stored in record body (removes duplication)
* Removes unnecessary decode in range
* Lots of misc bug fixes and exception fixes
* New Readme
* Added seek_one function
* Switched from UUID to bson.**ObjectId**

#### Version 0.1.14

* Switching to PyPi
* You can now download this package with "pip", note however it will only work with Python3 for now.

#### Version 0.1.4

* Fixed bug in empty to stop it deleting indecies
* Added 'save' method to table for updating pre-existing items (index aware)
* Added partial indexes, so only items where the key is not null are included in the index
* Code coverage hits 100%

#### Version 0.1.3

* Added seek(index, record) - seek to the first matching record based on an index key

  For example;
```  
table.index('by_compound', '{cat}|{name}')
for doc in table.seek('by_compound', {'cat': 'A', 'name': 'Squizzey'}):
    print(doc)
```

* Added range(index, lower, upper) - return all records with keys falling within the
  limits set up (upper, lower)

  For example;
  
```  
table.index('by_compound', '{cat}|{name}')
for doc in table.range('by_compound', {'cat': 'A', 'name': 'Squizzey'}, {'cat': 'B', 'name': 'Gareth Bult1'):
    print(doc)
```

#### Version 0.1.2

* Simplified the way indexes work, maintaining a second 'easy' method is no
  longer worthwhile with the new indexing interface. All indexes are now specified
  using format string, so a simple index on a field called name is specified as;
  
  - {name}
  
  A compound index based on name and age would be;
  
  - {name}{age}
  
  Appling a little logic, if you want records in age order, and age is always in
  the range 0-999 and is sorted numerically, we want the index specification to be;
  
  - {age:03}{name}
  
  i.e. a numberic, formatted as a string, length 3, zero padded
  
  If we wanted sorting by name, we would probably want;
  
  - {name}|{age:03}
  
  Not the use of a sepatating character (|) to ensure separation between the variable 
  length alphanumeric key and the fixed length numeric key.

#### Version 0.1.1

* Re-implemented 'find' method for Table
  * Now returns a generator rather than a list
  * Also accepts a filter, optionally applied after the index
  
* Implemented reindex method for Index class.
  Indexing a table containing data will create an initial index from this data.
  


examples/benchmarks.py
from pymamba import Database
from time import time
from random import random
from subprocess import call


def chunk(db, tbl, start, count):
    """
    Create a chunk of data
    
    :param tbl: Table to operate on
    :type tbl: Table
    :param start: Starting index 
    :type start: int
    :param count: Number of items
    :type count: int
    """
    begin = time()
    with db.begin():
        for index, session in enumerate(range(start, start + count)):
            rnd = random()
            record = {
	            'origin': 'linux.co.uk',
        	    'sid': start + count - index,
	            'when': time(),
        	    'day': int(rnd * 6),
            	'hour': int(rnd * 24)
        	}	
            tbl.append(record)

    finish = time()
    print("  - {:5}:{:5} - Append Speed/sec = {:.0f}".format(start, count, count / (finish - begin)))

print('** SINGLE Threaded benchmark **')
print('** Probably better throughput with multiple processes')
print('')

call(['rm', '-rf', 'databases/perfDB'])
print("* No Indecies")
db = Database('databases/perfDB')
table = db.table('sessions')
chunk(db,table, 0, 5000)
chunk(db,table, 5000, 5000)
chunk(db,table, 10000, 5000)
db.close()

call(['rm', '-rf', 'databases/perfDB'])
print("* Indexed by sid, day, hour")
db = Database('databases/perfDB')
table = db.table('sessions')
table.index('by_sid', '{sid}')
table.index('by_day', '{day}')
table.index('by_hour', '{hour}')
chunk(db,table, 0, 5000)
chunk(db,table, 5000, 5000)
chunk(db,table, 10000, 5000)
db.close()

call(['rm', '-rf', 'databases/perfDB'])
print("* Indexed by function")
db = Database('databases/perfDB')
table = db.table('sessions')
table.index('by_multiple', '!{origin}|{day:02}|{hour:02}|{sid:05}')
chunk(db, table, 0, 5000)
chunk(db,table, 5000, 5000)
chunk(db,table, 10000, 5000)

print("* Linear scan through most recent index")
start = 0
count = 15000
begin = time()
for doc in table.find('by_multiple'): pass
finish = time()
print("  - {:5}:{:5} - Read Speed/sec   = {:.0f}".format(start, count, count / (finish - begin)))

db.close()


